extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/AutoAddScopeBuildItem.java

```diff
@@ -0,0 +1,241 @@
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * This build item can be used to turn a class that is not annotated with a CDI scope annotation into a bean, i.e. the default
+ * scope annotation is added automatically if conditions are met.
+ */
+public final class AutoAddScopeBuildItem extends MultiBuildItem {
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private final MatchPredicate matchPredicate;
+    private final boolean injectable;
+    private final DotName defaultScope;
+    private final boolean unremovable;
+    private final String reason;
+
+    private AutoAddScopeBuildItem(MatchPredicate matchPredicate, boolean injectable,
+            DotName defaultScope, boolean unremovable, String reason) {
+        this.matchPredicate = matchPredicate;
+        this.injectable = injectable;
+        this.defaultScope = defaultScope;
+        this.unremovable = unremovable;
+        this.reason = reason;
+    }
+
+    public boolean isInjectable() {
+        return injectable;
+    }
+
+    public DotName getDefaultScope() {
+        return defaultScope;
+    }
+
+    public boolean isUnremovable() {
+        return unremovable;
+    }
+
+    public String getReason() {
+        return reason != null ? ": " + reason : "";
+    }
+
+    public boolean test(ClassInfo clazz, Collection<AnnotationInstance> annotations, IndexView index) {
+        return matchPredicate.test(clazz, annotations, index);
+    }
+
+    public interface MatchPredicate {
+
+        /**
+         * @param clazz
+         * @param annotations
+         * @param index
+         * @return {@code true} if the input arguments match the predicate,
+         *         otherwise {@code false}
+         */
+        boolean test(ClassInfo clazz, Collection<AnnotationInstance> annotations, IndexView index);
+
+        default MatchPredicate and(MatchPredicate other) {
+            return new MatchPredicate() {
+                @Override
+                public boolean test(ClassInfo clazz, Collection<AnnotationInstance> annotations, IndexView index) {
+                    return test(clazz, annotations, index) && other.test(clazz, annotations, index);
+                }
+            };
+        }
+
+    }
+
+    public static class Builder {
+
+        private MatchPredicate matchPredicate;
+        private boolean injectable;
+        private DotName defaultScope;
+        private boolean unremovable;
+        private String reason;
+
+        private Builder() {
+            this.defaultScope = BuiltinScope.DEPENDENT.getName();
+            this.unremovable = false;
+            this.injectable = false;
+        }
+
+        /**
+         * At least one injection point must be declared in the class hierarchy. Otherwise, the scope annotation is not added.
+         * 
+         * @return self
+         */
+        public Builder hasInjectionPoint() {
```

Comment:
It might be a stupid question and sorry if it's the case but should we also detect the lifecycle annotations? You might not have anything injected but a `@PostConstruct`?

Reply:
It's not a stupid question but I'm not sure all the specs support the lifecycle callbacks... In any case, we could add one more method to detect `@PostConstruct` and `@PreDestroy`.

Reply:
Well, that wouldn't hurt the ones that don't support it and would be nice for the ones that do? Or am I missing something?

Reply:
I get the point. We should rename the method then. It's only best effort anyway because even though we walk the class hierarchy we don't handle some special cases (e.g. an overriden initializer method)...

Reply:
Hm, what about `requiresContainerServices()`?